Support RRefs that contain threading.Locks

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57943 Support RRefs that contain threading.Locks**

This is a common scenario (our own tutorials propose it), hence we should ensure it works.

A more generic solution is desirable, but this should fix the immediate concern.

Differential Revision: [D28316076](https://our.internmc.facebook.com/intern/diff/D28316076/)

lw committed:
Support RRefs that contain threading.Locks

This is a common scenario (our own tutorials propose it), hence we should ensure it works.

A more generic solution is desirable, but this should fix the immediate concern.

Differential Revision: [D28316076](https://our.internmc.facebook.com/intern/diff/D28316076/)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -491,10 +491,12 @@ def inc_and_set(fut):
     def __init__(self, t):
         self.tensor = t
+        # Add one non-picklable field, to ensure it's ignored/skipped.
+        self.lock = Lock()
```
Shall we add a test to ensure this works?

lw replied:
Not sure I understand: this is the test. The method `test_cuda_future_can_extract_custom_class_with_cuda_tensor` will create an instance of this `TensorWrapper` class and pass it as a result to Future. By adding a lock as a field of TensorWrapper we would have caused a failure, but thanks to the fix this now works. Am I missing something?

mrshenli commented on torch/_jit_internal.py:
```diff
@@ -1132,8 +1137,15 @@ def persistent_id(self, obj):
+        # can just add a __getstate__ or __reduce__ method to their class.
+        if isinstance(obj, LockType):
+            return ""
+        return None
```
I have no objection against this change, but wonder whether JIT team is OK about it. One concern might be this class is called TensorExtractor, but it also cover locks.

lw replied:
Well it doesn't "cover" locks in the sense that it extracts locks too. It still just extracts tensors, it's just that now it explicitly knows that locks don't contain tensors and thus ignores them.

mrshenli reviewed (approved):
stamp to unblock

lw committed:
Update on "Support RRefs that contain threading.Locks"


This is a common scenario (our own tutorials propose it), hence we should ensure it works.

A more generic solution is desirable, but this should fix the immediate concern.

Differential Revision: [D28316076](https://our.internmc.facebook.com/intern/diff/D28316076/)

[ghstack-poisoned]

lw committed:
Update on "Support RRefs that contain threading.Locks"


This is a common scenario (our own tutorials propose it), hence we should ensure it works.

A more generic solution is desirable, but this should fix the immediate concern.

Differential Revision: [D28316076](https://our.internmc.facebook.com/intern/diff/D28316076/)

[ghstack-poisoned]

lw committed:
Update on "Support RRefs that contain threading.Locks"


This is a common scenario (our own tutorials propose it), hence we should ensure it works.

A more generic solution is desirable, but this should fix the immediate concern.

Differential Revision: [D28316076](https://our.internmc.facebook.com/intern/diff/D28316076/)

[ghstack-poisoned]

lw committed:
Update on "Support RRefs that contain threading.Locks"


This is a common scenario (our own tutorials propose it), hence we should ensure it works.

A more generic solution is desirable, but this should fix the immediate concern.

Differential Revision: [D28316076](https://our.internmc.facebook.com/intern/diff/D28316076/)

[ghstack-poisoned]

lw committed:
Update on "Support RRefs that contain threading.Locks"


This is a common scenario (our own tutorials propose it), hence we should ensure it works.

A more generic solution is desirable, but this should fix the immediate concern.

Differential Revision: [D28316076](https://our.internmc.facebook.com/intern/diff/D28316076/)

[ghstack-poisoned]

mrshenli commented:
Just to confirm, would I be correct if I assume the segfault we saw was fixed by #60470? Curious, where did we see the segfault? Is it in our unit tests or other applications?

mrshenli commented on torch/_jit_internal.py:
```diff
@@ -1154,8 +1172,19 @@ def persistent_id(self, obj):
+            return ""
+        # Futures and RRefs don't technically contain a value, they just offer
+        # the means to access a value.
+        if isinstance(obj, CFuture) or is_rref_instance(obj):
```
Trying to understand this. We don't care about Future and RRef even though they could contain tensors. Is it because we assume Future and RRef themselves should have already handle their Tensors properly?

lw replied:
I think at the end of the day this boils down to convention. [IValue::visit](https://github.com/pytorch/pytorch/blob/d3a8505ee1fbb79ba75962eea7891a0aaafc0b0a/aten/src/ATen/core/ivalue.cpp#L183-L184) and [IValue::getSubValues](https://github.com/pytorch/pytorch/blob/d3a8505ee1fbb79ba75962eea7891a0aaafc0b0a/aten/src/ATen/core/ivalue.cpp#L236-L241) don't "enter" Futures when they inspect them. Similarly if you take the future returned by `collectAll` and wait on it, it will _not_ automatically synchronize with the CUDA events of the "internal" futures: that must be done separately, manually.

Given all that it feels natural and consistent here to also not "enter" Futures and RRefs when extracting tensors.

lw commented:
Yes the issue that popped up after this PR is being fixed in #60470. It consisted in accessing a deleted object, so it could manifest as a segfault or just as "memory corruption" (reading invalid data). In my case, the error I got looked a lot like this one: #60440. I was hitting it when running some code written by @pbelevich, but note that in #60470 I also added a test which triggered that.

mrshenli reviewed (approved):
Stamp to unblock. 

Shall we add a test to cover modules with RRef and locks?

lw committed:
Update on "Support RRefs that contain threading.Locks"


This is a common scenario (our own tutorials propose it), hence we should ensure it works.

A more generic solution is desirable, but this should fix the immediate concern.

Differential Revision: [D28316076](https://our.internmc.facebook.com/intern/diff/D28316076/)

[ghstack-poisoned]
